### PR TITLE
feat(csharp/backtracking): add csharp code in backtracking

### PR DESCRIPTION
**Pull request by hpstory**
If this PR is related to coding or code translation, please read the [contribution guideline](https://github.com/krahets/hello-algo/issues/15#issue-1464069591), fill out the checklist, and paste the console outputs to the PR.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

测试结果：
backtrack_find_constrained_paths
preorder_find_constrained_paths
preorder_find_nodes
Screenshot 2023-04-17 113023

### CODE REVIEW

File: codes/csharp/chapter_backtracking/preorder_find_constrained_paths.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        preOrder(root.left);
+        preOrder(root.right);
+        // 回退
+        path.RemoveAt(path.Count - 1);
+    }
+
+    [Test]
+    public void Test()
+    {
+        TreeNode root = TreeNode.ArrToTree(new int?[] { 1, 7, 3, 4, 5, 6, 7 });
```
Comment (hpstory) [inline codes/csharp/chapter_backtracking/preorder_find_constrained_paths.cs:41]: Can I change this method(ArrToTree -> listToTree) to be the same signature as other languages? If possible, I will create a new PR to modify it.

Comment (krahets) [reply codes/csharp/chapter_backtracking/preorder_find_constrained_paths.cs:41]: LGTM! Let's change it.

Comment (krahets) [reply codes/csharp/chapter_backtracking/preorder_find_constrained_paths.cs:41]: Don't hesitate to introduce more ideas. Thanks!

**Review by krahets — APPROVED** [review]
Thanks!

**Pull request merged**